core/src/main/java/com/microsoft/applicationinsights/internal/channel/common/TelemetryBuffer.java

```diff
@@ -55,25 +57,32 @@
      */
     private final class TelemetryBufferTelemetriesFetcher implements TelemetriesTransmitter.TelemetriesFetcher {
 
-        private final long expectedGeneration;
+        private final AtomicLong expectedGeneration;
 
-        private TelemetryBufferTelemetriesFetcher(long expectedGeneration) {
+        private TelemetryBufferTelemetriesFetcher(AtomicLong expectedGeneration) {
             this.expectedGeneration = expectedGeneration;
         }
 
+        // No need for low-level synchronization. Atomic long cover cover the generation check consistency
+        // ConcurrentLinkedList handles other parts for collection operation.
         @Override
         public Collection<T> fetch() {
-            synchronized (lock) {
-                if (expectedGeneration != generation) {
-                    return Collections.emptyList();
-                }
-
-                ++generation;
-                List<T> readyToBeSent = telemetries;
-                telemetries = new ArrayList<T>();
 
-                return readyToBeSent;
+            // cannot use equals() method here as AtomicLong doesn't
+            // override equals and hash code
+            if (expectedGeneration.get() != generation.get()) {
+                return Collections.emptyList();
             }
+
+            generation.incrementAndGet();
+
+            // Create an unmodifiable view of the collection so that the items cannot be edited
+            Collection<T> readyToBeSent = new ArrayList<>(Collections.unmodifiableCollection(telemetries));
+
```

Comment:
This doesn't seem to be the correct way of using concurrent collections.

Reply:
You don't need unmodifiableCollection here.